README
KGraphViewer is a Graphviz dot graph file viewer for KDE.
https://gna.org/projects/kgraphviewer/
Copyright (C) 2005-2006 Gaël de Chalendar <kleag@free.fr>
    
    KGraphViewer is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation, version 2.
    
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    General Public License for more details.
    
    You should have received a copy of the GNU General Public License
    along with this program; see the file COPYING.  If not, write to
    the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
    Boston, MA 02110-1301, USA.


I have created a dot graphs (Graphviz package, see http://www.graphviz.org/)
for KDE. I used, in this program, code from the profiling tool kcachegrind
 (http://kcachegrind.sourceforge.net/cgi-bin/show.cgi/KcacheGrindIndex).
For printing, I reused code from the KDE project koffice, namely Kexi, the
database editor.

The tool is not finished but already usable for not huge graphs (< 1000 nodes).
Here are its currently implemented features:
- clusters handling
- zooming
- loading of several graphs in tabs;
- saving of the recent files list;
- manual reload of files;
- display of a bird-eye view of the graph;
- moving of the graph by dragging;
- simple graph printing with preview (no resizing nor cutting in pages);
- closing of the tabs;
- loading of nearly all graphviz example graphs;
- nodes and edges colors and all node shapes are supported

Kgraphviewer is using the boost/spirit library for graphs parsing. To avoid
external dependencies, these Boost parts have been integrated in Kgraphviewer.

See TODO in the distribution for planned features.

Kgraphviewer is currently only available as source from its subversion
repository. See https://gna.org/svn/?group=kgraphviewer for instructions.

As usual, comments, bug reports and any help are very welcome !

Please report bugs or install problems directly on the kgraphviewer homepage
and not as comments here. Thanks.

Kleag


ChangeLog
2014/03/29
  * compilation against latest graphviz
    To compile against versions before 2.30 you need to
    manually apply old_graphviz.diff first.
  * switch to KPluginFactory
  * fix scrolling through graph with cursor keys
  * many small fixes of UI bugs and annoyances

2010/10/10
  * Set version to 2.1
  * Bug corrected: edges line width and color and nodes line width correctly handled

2010/09/19
  * new background color option

2008/05/19
  * version 2.0.2: works with boost-1.34


========= VERY LONG PAUSE IN THE ChangeLog=================

2006/09/20
  * X11 colors names now unsupported by Qt are harcoded in KGraphViewer

2006/09/18
  * dcop -> dbus port finished (thanks to the help on #kde4-devel)

2006/09/17
  * graph tips (on nodes and edges)

2006/09/16
  * KSelectActions now work as before
  
2006/09/14
  * asked and obtained the removing of -D QT_NO_STL from default compile flags
  * solved with help from IRC the kintnuminput.h -> knuminput.h problem in file 
    generated from ui

2006/07/02
  * fixed a xdot parsing bug
  * fixed a crash condition
  * better handling of parse errors: shows a dialog 

2006/05/27
  * removed a bug appearing with graphviz 2.8

2006/05/26
  * addad in the menu bar the entries from contextual menu and reported this 
    change in documentation
  * added the possibility to disable the bird-eye view (menu, context menu, 
    accel, toolbar) and reported this change in documentation
  * check for memory leaks

2006/05/25
  * added a font cache to optimize a little bit

2006/05/24
  * changes in config dialog now change the status of Apply button
  * make appear the bird-eye view even if it was not initialy appearent when 
    zooming
  * optimizations to be able to load and manipulate larger graphs (200+ linear 
    graph OK)

2006/05/23
  * documentation

2006/05/18
  * corrected placing of window position indicator (red line) in bird eye view 
    when zooming
  * add tooltips and whatsthese everywhere
  * corrected tooltips placement
  * added French translation

2006/05/17
  * corrected drawing errors appearing when dragging after a zoom by updating
    Qt to 3.3.6

2006/05/16
  * all graphviz example graphs are perfectly drawn
  * corrected drawing of parallel edges
  * corrected edges geometry definition (tips appear at wrong places)
  * allow opening of multiple files at the same time

2006/05/15
  * corrected loading of edges bounds labels
  * correct handling of record shaped nodes (see graphviz examples, compare with 
    doted results)
  * corrected handling of fonts (see graphviz examples, compare with 
    doted results)
  * corrected drawing of some undirected graphs in graphviz examples (compare with 
    doted results)
  * added support for main graph label

2006/05/14
  * corrected drawing of clusters (compared with doted results)
  * hack to not freeze with huge graphs. Stop loading after N vertices or edges
    (N currently set to 1000)

2006/05/13
  * maked the window position indicator to be still visible even when zooming a 
    lot (min size or big bold point)

2006/05/12
  * added hard-coded min and max zoom factors
  * scrolling with mouse wheel
  * make minimum canvas size to window size
  * reduce canvas size to graph size when its view becomes larger than itself

2006/05/07
  * better integration of printing (splitting in pages, etc.)
  * change reload icon to a "recycling" icon

2006/05/07
  * printing on multiple pages or fited to one page
  * applying the scaling factor to lines width (since more time but not logged)

2006/05/05
  * added print preview and setup (code reused from the koffice tool kexi)

2006/04/15
  * change mouse pointer to grabbing hand when scrolling

2006/04/13
  * Support for Graphviz 2.8

2006/04/12
  * Bug corrected: crash when closing last tab ;
  * First implementation of support for the URL attribute: the focus follows the
    mouse and clicking on it opens the default browse with the URL (only fully
    qualified URLs are supported for now, not relative paths to files) ;
  * Bug correction: the label is now drawn over the image if any ;

2006/04/11
  * added support for a new dot language attribut: shapefile (node defined by
    an image)
  
2006/04/08
  * session management (reload of previously opened files)
  * context menu to use other layouting algorithms (twopi, fdp, etc.)
  * added a toolbar entry to choose layouting program

2006/03/31
  * automatic reloading (with user confirmation or not) of (externaly)  
    modified files ;

2006/03/30
  * open new instances as new tabs in the old one (should be configurable)

2006/03/29
  * menus and toolbars reorganisation
  * empty file names no more stored in recent files list
  * loading of parallel edges
  * automaticaly choose dot for directed graphs and neato for undirected
  * handling of comments in dot files

2006/03/22
  * important changes. A lot of graphs are now perfectly diplayed.
  * UTF-8 chars in labels now correctly handled.
  
2005/05/20
  * first step towards huge graphs loading. 800 (configurable in code)
    nodes are displayed at max. A lot of work remains on this subject.

2005/05/11
  * add support to all nodes shapes ;
  * correct fill color handling ;
  * parsing without errors of all (?) valid dot files ;
  * added boost/spirit (from boost 1.32) and helpers with appropriate 
    copyright and licensing information in the project to avoid external 
    dependencies

2005/05/07
  * nearly all dot examples load, some of them perfectly (remains 1 
    encoding problem, 1 infinite loop and one graph not handled by dot)

2005/04/28
  * using Boost/Spirit parser generator for parsing dot output
  * handling of fonts in labels



src/main.cpp
/* This file is part of KGraphViewer.
   Copyright (C) 2005-2010 Gael de Chalendar <kleag@free.fr>

   KGraphViewer is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public
   License as published by the Free Software Foundation, version 2.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
   02110-1301, USA
*/


#include "kgraphviewer.h"
#include <kuniqueapplication.h>
#include <kaboutdata.h>
#include <kcmdlineargs.h>
#include <kmessagebox.h>
#include <klocale.h>
#include <kdebug.h>
#include <iostream>
#include <qdir.h>
#include <QByteArray>
#include <QDBusInterface>
#include <QDBusConnection>
#include <QDBusConnectionInterface>
#include <QDBusReply>

#include "kgraphvieweradaptor.h"

static const char description[] =
I18N_NOOP("A Graphviz dot graph viewer for KDE");

static const char version[] = "2.1.1";

int main(int argc, char **argv)
{
  KAboutData about("kgraphviewer", 0, ki18n("KGraphViewer"), version, ki18n(description),
                    KAboutData::License_GPL, ki18n("(C) 2005-2010 Gaël de Chalendar"), KLocalizedString(), 0, "kleag@free.fr");
  about.addAuthor( ki18n("Gaël de Chalendar"), ki18n("Original Author and current maintainer"), "kleag@free.fr" );
  about.addAuthor( ki18n("Reimar Döffinger"), ki18n("Contributor"), "kde@reimardoeffinger.de" );
  about.addAuthor( ki18n("Matthias Peinhardt"), ki18n("Contributor"), "matthias.peinhardt@googlemail.com" );
  about.addAuthor( ki18n("Sandro Andrade"), ki18n("Contributor"), "sandro.andrade@gmail.com" );
  about.addAuthor( ki18n("Milian Wolff"), ki18n("Contributor"), "mail@milianw.de" );
  
  KCmdLineArgs::init(argc, argv, &about);

  KCmdLineOptions options;
  options.add("+[URL]", ki18n( "Dot graph to open" ));
  KCmdLineArgs::addCmdLineOptions( options );
  KApplication app;
  
// see if we are starting with session management
  if (app.isSessionRestored())
  {
    RESTORE(KGraphViewerWindow);
  }
  else
  {
      // no session.. just start up normally
      KCmdLineArgs *args = KCmdLineArgs::parsedArgs();

      KGraphViewerWindow *widget = 0;
      if ( args->count() == 0 )
      {
        widget = new KGraphViewerWindow;
        new KgraphviewerAdaptor(widget);
        QDBusConnection::sessionBus().registerObject("/KGraphViewer", widget);
        widget->show();
      }
      else
      {
        QDBusReply<bool> reply = QDBusConnection::sessionBus().interface()->isServiceRegistered( "org.kde.kgraphviewer" );
      
        bool instanceExists = reply.value();
  
        for (int i = 0; i < args->count(); i++ )
        {
          if (instanceExists 
              && (KMessageBox::questionYesNo(0, 
                                         i18n("A KGraphViewer window is already open, where do you want to open this file?"),
                                         i18n("Opening in new window confirmation"),
                                             KGuiItem(i18nc("Where to open a file", "Existing window")),
                                             KGuiItem(i18nc("Where to open a file", "New window")),
                                             "openInNewWindowMode"   ) == KMessageBox::Yes) )
          {
            QByteArray tosenddata;
            QDataStream arg(&tosenddata, QIODevice::WriteOnly);
            QString strarg = args->arg(i);
            KUrl url;
            if (strarg.left(1) == "/")
              url = KUrl(strarg);
            else url = KUrl(QDir::currentPath() + '/' + strarg);
            arg << url;
            QDBusInterface iface("org.kde.kgraphviewer", "/KGraphViewer", "", QDBusConnection::sessionBus());
            if (iface.isValid()) 
            {
              QDBusReply<void> reply = iface.call("openUrl", url.pathOrUrl());
              if (reply.isValid()) 
              {
                kDebug() << "Reply was valid";
                return 0;
              }

              kError() << "Call failed: " << reply.error().message() << endl;
              return 1;
            }
            kError() << "Invalid interface" << endl;
            exit(0);
          }
          else
          {
            widget = new KGraphViewerWindow;
            new KgraphviewerAdaptor(widget);
            QDBusConnection::sessionBus().registerObject("/KGraphViewer", widget);
            widget->show();
            widget->openUrl( args->url( i ) );
          }
        }
      }
      args->clear();
    if (widget != 0)
    {
      widget->  reloadPreviousFiles();
    }
    
  }
  return app.exec();
}
